### ISSUE TITLE
Phase 3b: TLS-fingerprint ML scoring (signal, not gate)

### ISSUE THREAD

**Issue by fabriziosalmi** [fabriziosalmi/zion#28]
A small supervised classifier scores every incoming connection's TLS fingerprint and forwards the result as headers (`X-Client-TLS-Score`, `X-Client-TLS-Class`). Zion never blocks based on the score — the upstream application decides what to do with it. Same architectural decision Cloudflare made with their commercial Bot Score: signal, not gate.

This makes Zion the first OSS edge gateway to ship a Cloudflare-Bot-Score-style classifier as a built-in feature, rather than as a paid add-on.

## Why this complements Phase 3a

| | Phase 3a (#27) | Phase 3b (this issue) |
|---|---|---|
| Output | sì / no | continuous score ∈ [0, 1] |
| Match type | exact JA4 lookup | feature-vector classification |
| Use case | "I distribute the client to my workers" | "I serve the public web, I want to grade incoming traffic" |
| Default | `mode = "off"` (deny-by-default when on) | `mode = "score"` (forward header, no block) |
| Bot adversary | strong (must replicate exact JA4) | strict subset of fingerprints adversary must replicate |
| New version of Chrome | manual allowlist update | model generalises ("Chrome-like" pattern) |

The two features are NOT alternatives — a deployment can run Phase 3a allowlist for known workers and ML score for the rest of public traffic, on the same `[tls.fingerprint]` section.

## Architecture (superlight, by design)

The classifier runs INLINE in the request pipeline. Constraints:
* Inference cost ≤ 1 µs per request (i.e. < 1 % of TLS handshake cost).
* No GPU. No TensorFlow. No PyTorch runtime in the binary.
* Explainable: a SOC analyst must be able to read why a request scored 0.8.

Selected stack:
* Model: **logistic regression** with ~50 features. Training in Python (`scikit-learn`), inference baked into the Rust binary as a `[f64; N]` weight array.
* No ML runtime dependency. The model IS the binary. Refresh = new release.
* Phase 3c (#29) introduces ONNX runtime for customers who want their own model.

Feature vector (~50–80 dimensions extracted from the ClientHello):
* TLS version (one-hot)
* Cipher suite presence vector + order positions of top-N ciphers
* Extension presence vector + order positions
* `supported_groups`, `signature_algorithms`, `ec_point_formats`, `compression_methods`
* ALPN list + order
* GREASE values position (Google's anti-fingerprint canary, position is discriminating)
* `padding_length`, total record bytes
* Has `application_settings`? (Chrome-only marker)
* Has `encrypted_client_hello`?

Hot path:
```text
peek ClientHello (~500 ns)
  → tls-parser → feature vector  (~500 ns)
  → JA4 compute                    (~500 ns)
  → model.predict(features)        (~100 ns logistic regression)
  → metric tls_score_bucket{class="bot"}++   (~30 ns)
  → forward as X-Client-TLS-Score / X-Client-TLS-Class
  → rustls handshake proceeds normally (peeked bytes still in kernel buffer)

Total added per request: ~2 µs ≈ 1 % of TLS handshake cost.
```

## Header schema

All forwarded headers, all forwarded as opt-in via `forward_*` flags:

```
X-Client-TLS-JA4:      t13d1516h2_8daaf6152771_b186095e22b6  (already shipped in Phase 3a)
X-Client-TLS-Class:    browser | tool | bot | unknown        (top-1 prediction)
X-Client-TLS-Score:    0.83                                   (P(bot) ∈ [0, 1])
X-Client-TLS-Model:    builtin-v1                             (for audit / debugging)
```

Inbound-header strip (`X-Client-TLS-*` arriving from outside) is enforced exactly like `X-Client-Cert-Fingerprint` and `X-Real-IP` are stripped today (since v0.1.7).

## Decision rule, downstream

Zion exposes the score; the application decides. Example pattern (in any language):

```js
// Express-style middleware. Zion's signals here, app's policy here.
function gate(req, res, next) {
  const score = parseFloat(req.headers['x-client-tls-score'] ?? '0');
  const klass = req.headers['x-client-tls-class'] ?? 'unknown';
  const userId = req.session?.userId;

  if (userId && score > 0.7) return requireMfa(req, res);                 // step-up
  if (!userId && score > 0.5 && req.method !== 'GET') return showCaptcha();// public + write
  if (req.path.startsWith('/admin') && klass === 'unknown') return res.status(403).end();
  next();
}
```

Same Zion deployment, two apps, two policies. Zion stays observer.

## Config shape

```toml
[tls.fingerprint]
mode = "score"               # off | shadow | score | allowlist
                              # off:       no peek
                              # shadow:    peek + JA4 + log + counter, no header
                              # score:     peek + JA4 + ML score + forward as header
                              # allowlist: see #27 (Phase 3a)

# `score` mode parameters:
model = "builtin-v1"         # builtin-v1 (baked) | onnx:/etc/zion/bot.onnx (Phase 3c)

forward_score   = true
forward_class   = true
forward_ja4     = true
log_threshold   = 0.5         # log every request above this score
log_full_hello  = false       # privacy: opt-in only
```

## Roadmap (incremental commits)

1. **Feature extractor**: `src/tls_fp.rs` (or extension of Phase 3a's module): parse ClientHello → feature vector. Stable serialisation. Unit tests against canned ClientHello samples (Chrome 130, Firefox 134, Safari 18, curl 8.4, python-requests 2.32, Go net/http).
2. **Training pipeline** (`scripts/train-tls-classifier/`): Python sklearn against CIC-IDS-2017/2018 + CTU-13 datasets. Output: `model.json` with `{coefficients: [f64; N], feature_names: [...], threshold: 0.5, version: "builtin-v1"}`. Build script reads it and codegens `src/tls_fp_model.rs`.
3. **Inference**: pure Rust, ~30 LoC, no deps. Sigmoid + dot product.
4. **Header forwarding** + Prometheus counters with **bounded cardinality** (top-N JA4 buckets, not unique-JA4 cardinality bomb).
5. Docs (`docs/security/tls-scoring.md`) covering: how to retrain on customer data, drift detection, explainability of weights, adversarial robustness limits.

Total estimated diff: ~600 LoC Rust + ~500 LoC Python (training scripts), no new runtime dependencies.

## Acceptance criteria

* [ ] `mode = "off"` → zero peek, zero overhead.
* [ ] `mode = "shadow"` → JA4 + class predicted + counter; NO header forwarded; NO upstream visibility into the score.
* [ ] `mode = "score"` → score and class headers forwarded to upstream. Application can read them.
* [ ] Inbound `X-Client-TLS-Score` / `X-Client-TLS-Class` from outside is stripped before forwarding (test: spoof attempt → header absent at upstream).
* [ ] Per-class bucketed counter `tls_score_total{class="browser|tool|bot|unknown"}` exposed in `/metrics`.
* [ ] Inference cost measured: < 5 µs P99 over 100K samples on M4 / equivalent.
* [ ] Model identifier (`zion_tls_model_version` Prometheus gauge) exposed so dashboards can detect "old model still running".
* [ ] Training pipeline reproducible from `scripts/train-tls-classifier/README.md` — public dataset references, expected output dimensions, accuracy benchmarks.
* [ ] Score never used as a gate. No code path inside Zion blocks based on the score. **Documentation explicit on this.**

## Risks and how we mitigate them

1. **Adversarial robustness.** A logistic regression baked into an OSS binary is fully transparent to the adversary. Mitigation: this is a *filter*, not a *defense*. Phase 3b catches the long tail (Python `requests`, Go default, curl) that uses default TLS stacks; sophisticated adversaries (`curl-impersonate`) bypass it but they are also rare and visible in other ways.
2. **Distribution drift.** TLS landscape evolves (TLS 1.3 ECH, post-quantum cipher rollout). Mitigation: model version pinned in metric, `model_version` field in `/_zion/snapshot.json`, explicit retrain cadence documented (every minor release of Zion).
3. **Dataset bias.** CIC-IDS is US-skewed. Mitigation: documented limitation; customers with non-US traffic can retrain on their own corpus via Phase 3c (#29).
4. **Explainability for compliance.** Logistic regression weights are *the* explanation. SHAP values are exposable as a debug endpoint if a regulator asks "why was this request flagged?". Mitigation: pin logistic regression as the model class; do NOT silently switch to a deep network in a future PR without a corresponding explainability story.
5. **False positives on the long tail of legit traffic** (Brave, Tor, hardened Firefox). Mitigation: by design, Zion never blocks. The score is an upstream-app decision; the app can whitelist its known weird clients.

## Out of scope (this issue)

* Loading customer-provided ONNX models — see #29 (Phase 3c).
* Online learning / drift adaptation — manual retrain only.
* Dataset distribution / hosting — point to public datasets, no hosting in repo.
* GPU-based inference — not in the cards for an OSS edge gateway.

## References

* Cloudflare Bot Score (architectural inspiration, closed-source comparator)
* CIC-IDS-2017/2018 dataset: https://www.unb.ca/cic/datasets/ids-2017.html
* CTU-13 botnet dataset: https://www.stratosphereips.org/datasets-ctu13
* JA4 spec: https://github.com/FoxIO-LLC/ja4
* Phase 3a (allowlist): #27
* Phase 3c (ONNX runtime): #29
